mymake.cpp

```diff
@@ -218,14 +226,45 @@ int main(int argc, char **argv) {
       }
     time_t obj_time = get_file_time(obj);
     if(src_time > obj_time) {
-      printf("compiling %s... [%d/%d]\n", m.c_str(), id, int(modules.size()));
-      if(system(compiler + " " + opts + " " + src + " -o " + obj)) { printf("error\n"); exit(1); }
+      pair<int, string> task(id, compiler + " " + opts + " " + src + " -o " + obj);
+      tasks.push_back(task);
       }
     else {
       printf("ok: %s\n", m.c_str());
       }
     allobj += " ";
     allobj += obj;
+    id++;
+    }
+
+  chrono::milliseconds quantum(25);
+  vector<future<int>> workers(batch_size);
+
+  int tasks_amt = tasks.size();
+  int tasks_taken = 0, tasks_done = 0;
+  bool finished = tasks.empty();
+
+  while (!finished)
+  for (auto & worker : workers) {
+    check_lollygagging:
+    if (worker.valid()) {
+      if (worker.wait_for(quantum) != future_status::ready) continue;
+      else {
+        int res = worker.get();
+        if (res) { printf("compilation error!\n"); exit(1); }
+        ++tasks_done;
+        goto check_lollygagging;
+        }
+      }
+    else if (tasks_taken < tasks_amt) {
+      auto task = tasks[tasks_taken];
+      int mid = task.first;
+      string cmdline = task.second;
+      printf("compiling %s... [%d/%d]\n", modules[mid].c_str(), tasks_taken+1, tasks_amt);
+      worker = async(launch::async, (int (*)(string))system, cmdline);
```

Comment:
I recommend making `tasks` a vector of `std::function<void(int, int)>`; then replace these 5 lines with
```
worker = std::async(std::launch::async, tasks[tasks_taken], tasks_taken+1, tasks_amt);
```
and replace lines 229–230 with
```
std::string cmdline = compiler + " " + opts + " " + src + " -o " + obj;
tasks.push_back([m, cmdline](int x, int y) {
    printf("compiling %s... [%d/%d]\n", m.c_str(), x, y);
    system(cmdline);
});
```
This eliminates the UB in `(int (*)(string))system` (namely, you're not allowed to assume that the standard library function has an overload with that exact signature).

Packaging up the `system` call into a `std::function` is the main thing. I admit it might be cleaner to keep the module name `m` separately outside the function (i.e. `std::pair<std::string, std::function<void()>>`), rather than passing `tasks_taken` and `tasks_amt` back _into_ the function.

Reply:
Okay, that conversion did indeed feel quite ugly to me, however in this case `system(string)` is a locally-defined function, so (correct me if that's wrong) I guess `system` can be assumed to have this overload?
https://github.com/zenorogue/hyperrogue/blob/4386f86efb88c065fd82b3bf42fc8ba8f88d3eb0/mymake.cpp#L73-L75

I'll probably need some more time to process this suggestion. It does seem more elegant, though it needs at least a minor change because the return value of `system` call is of interest. And I don't imagine that can be packaged into a lambda (as a simple `printf` message, at least) because right now the behavior is that once one module fails compilation, everything stops; it feels like a good idea to keep this behavior.

Actually, if we are to go the route of packing the compilation-notification message into a lambda, why not just pre-bake the string-to-be-`printf`d?

Reply:
I thought about this for a bit and decided against packing messages into `std::function`. So that a `task` is a `std::pair<int, std::function<int(void)>>`. I'll commit the changes shortly.